addons/project_timesheet_holidays/models/resource_calendar_leaves.py

```diff
@@ -151,15 +151,16 @@ def create(self, vals_list):
         return results
 
     def write(self, vals):
-        start_dates = self.mapped('date_from')
-        end_dates = self.mapped('date_to')
+        date_from, date_to, calendar_id = vals.get('date_from'), vals.get('date_to'), vals.get('calendar_id')
+        global_time_off_updated = self.env['resource.calendar.leaves']
+        if date_from or date_to or 'calendar_id' in vals:
+            global_time_off_updated = self.filtered(lambda r: (date_from is not None and r.date_from != date_from) or (date_to is not None and r.date_to != date_to) or (calendar_id is not None and r.calendar_id.id != calendar_id))
+            timesheets = global_time_off_updated.timesheet_ids
+            if timesheets:
+                timesheets.write({'global_leave_id': False})
+                timesheets.unlink()
         result = super(ResourceCalendarLeaves, self).write(vals)
-        date_from, date_to = vals.get('date_from'), vals.get('date_to')
-        if date_from or date_to:
-            if any(start_date != date_from or end_date != date_to for start_date, end_date in zip(start_dates, end_dates)):
-                timesheets = self.mapped('timesheet_ids')
-                if timesheets:
-                    timesheets.write({'global_leave_id': False})
-                    timesheets.unlink()
-                    self._timesheet_create_lines()
+        if global_time_off_updated:
+            global_time_offs_with_leave_timesheet = global_time_off_updated.filtered(lambda r: not r.resource_id and r.calendar_id.company_id.internal_project_id and r.calendar_id.company_id.leave_timesheet_task_id)
```

Comment:
Can you check if when the user has not access to the project, we have no access error when we try to check if the company linked to the working schedule is linked to a project? I think we probably have to add a sudo.
```suggestion
            global_time_offs_with_leave_timesheet = global_time_off_updated.filtered(lambda r: not r.resource_id and r.calendar_id.company_id.sudo().internal_project_id and r.calendar_id.company_id.sudo().leave_timesheet_task_id)
```

Also why not using the `company_id` field defined in the `resource.calendar.leave`? 🤔

Reply:
Isn't `internal_project_id` a project that is always accessible for every employees? I inspired myself from the create function of `resource.calendar.leave` so there is a problem there too if there is one here :thinking:

Reply:
I am not sure that's why I ask to check it to be sure to not have any access error. 🙂 
Sorry I don't understand the second part of your message. 🤔

Reply:
Indeed you are right, it is always accessible since we just check if the field is filled in the company.